event-dispatcher/src/main/java/io/sk8s/event/dispatcher/EventDispatchingHandler.java

```diff
@@ -16,55 +16,264 @@
+
+	private volatile long scalingFrequencyMs = 0L; // Wait forever
+
+	private final Thread scalingThread = new Thread(new ScalingMonitor());
+
+	private volatile boolean running = false;
+
+	private final KafkaConsumerMonitor kafkaConsumerMonitor = new KafkaConsumerMonitor(
+			System.getenv("SPRING_CLOUD_STREAM_KAFKA_BINDER_BROKERS"));
 
 	@EventListener
-	public void onFunctionRegistered(ResourceAddedEvent<XFunction> event) {
+	public synchronized void onFunctionRegistered(ResourceAddedOrModifiedEvent<XFunction> event) {
 		XFunction functionResource = event.getResource();
 		String functionName = functionResource.getMetadata().getName();
 		this.functions.put(functionName, functionResource);
-		this.launcher.dispatch(functionResource);
+		this.kafkaConsumerMonitor.beginTracking(functionName, functionResource.getSpec().getInput());
+
+		updateWakeUpInterval();
+		if (!running) {
+			running = true;
+			scalingThread.start();
+		}
+		this.notify();
+
 		logger.info("function added: " + functionName);
 	}
 
 	@EventListener
-	public void onFunctionUnregistered(ResourceDeletedEvent<XFunction> event) {
+	public synchronized void onFunctionUnregistered(ResourceDeletedEvent<XFunction> event) {
 		XFunction functionResource = event.getResource();
 		String functionName = functionResource.getMetadata().getName();
-		this.kubernetesClient.extensions().jobs().withLabel("function", functionName).delete();
 		this.functions.remove(functionName);
+		this.actualNbReplicas.remove(functionName);
+
+		deployer.undeploy(functionResource);
+
+		this.kafkaConsumerMonitor.stopTracking(functionName, functionResource.getSpec().getInput());
+
+		updateWakeUpInterval();
+		this.notify();
+
 		logger.info("function deleted: " + functionName);
 	}
+
+	@EventListener
+	public void onTopicRegistered(ResourceAddedOrModifiedEvent<Topic> event) {
+		Topic topic = event.getResource();
+		String name = topic.getMetadata().getName();
+		topics.put(name, topic);
+		topicsReady.computeIfAbsent(name, i -> new CountDownLatch(1)).countDown();
+	}
+
+	@EventListener
+	public void onTopicUnregistered(ResourceDeletedEvent<Topic> event) {
+		Topic topic = event.getResource();
+		String name = topic.getMetadata().getName();
+		topics.remove(name);
+		topicsReady.remove(name);
+	}
+
+	@EventListener
+	public void onDeploymentRegistered(ResourceAddedOrModifiedEvent<Deployment> event) {
+		Deployment deployment = event.getResource();
+		String functionName = deployment.getMetadata().getName();
+		actualNbReplicas.put(functionName, deployment.getSpec().getReplicas());
+	}
+
+	@EventListener
+	public void onDeploymentUnregistered(ResourceDeletedEvent<Deployment> event) {
+		Deployment deployment = event.getResource();
+		String functionName = deployment.getMetadata().getName();
+		actualNbReplicas.remove(functionName);
+	}
+
+	@PreDestroy
+	public synchronized void tearDown() {
+		running = false;
+		this.notify();
+		kafkaConsumerMonitor.close();
+	}
+
+	private void updateWakeUpInterval() {
```

Comment:
might as well pick one name consistently: wakeup or scaling? (I prefer the latter)

Reply:
Not exactly the same. The thread wakes up at said interval, but does not mean that it scales (as in "changes the value") on each wakeup. Still want `updateScalingInterval`?

Reply:
what I mean is that inside this method the actual update applies to a var called "scalingFrequencyMs". I was suggesting that the method should describe what it's updating more accurately. I would probably call it the "scalingMonitorInterval" (that way it's not suggesting that it will actually *scale* just *monitor*, and as I pointed out elsewhere it's technically an interval and not a frequency, which would be N-per-some-timeunit)